File: apps/web/src/types/risk.ts
export type RiskDisplayDecision = "PASS" | "REJECT" | "REVIEW";

export interface RiskRuleResult {
  id: string;
  seq: number;
  rule_key: string;
  decision: "ALLOW" | "REJECT" | "REQUIRE_CONFIRMATION";
  reason_code: string;
  message: string;
  severity: string;
  observed_value: string | number | boolean | null;
  limit_value: string | number | boolean | null;
  evaluated_at: string;
}

export interface RiskDecision {
  id: string;
  request_id: string;
  source_type: string;
  source_id: string | null;
  account_id: string;
  instrument_id: string;
  account: { id: string; code: string | null; name: string | null };
  instrument: {
    id: string;
    symbol: string | null;
    exchange: string | null;
    name: string | null;
  };
  overall_decision: "ALLOW" | "REJECT" | "REQUIRE_CONFIRMATION";
  side: string | null;
  order_type: string | null;
  quantity: string | number | null;
  estimated_notional: string | null;
  projected_instrument_weight: string | null;
  projected_total_exposure: string | null;
  limits_snapshot: Record<string, unknown>;
  account_snapshot: Record<string, unknown>;
  instrument_snapshot: Record<string, unknown>;
  warnings: string[];
  rule_results: RiskRuleResult[];
  risk_rule_summary: Array<Record<string, unknown>>;
  order_id: string | null;
  correlation_id: string;
  evaluated_at: string;
}

export interface RiskDecisionPage {
  items: RiskDecision[];
  page: number;
  page_size: number;
  total: number;
}

export interface ActiveRiskLimits {
  max_order_notional: string | null;
  max_instrument_weight: string | null;
  max_total_exposure: string | null;
  max_orders_per_window: number | null;
  order_frequency_window_seconds: number;
  allow_market_orders: boolean;
  require_reference_price_for_market_order: boolean;
  kill_switch_enabled: boolean;
  configuration_source: string;
  effective_at: string;
  warnings: string[];
}

File: apps/api/src/alphadesk_api/schemas/risk.py
"""Read-only R01 risk API contracts."""

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field, field_validator


class RiskDecisionResponse(BaseModel):
    model_config = {"extra": "allow"}

    id: str
    idempotency_key: str
    request_fingerprint: str
    request_id: str
    source_type: str
    account_id: str
    instrument_id: str
    overall_decision: str
    order_id: str | None
    warnings: list[str]
    rule_results: list[dict[str, Any]]


class ActiveRiskLimitsResponse(BaseModel):
    max_order_notional: str | None
    max_instrument_weight: str | None
    max_total_exposure: str | None
    max_orders_per_window: int | None
    order_frequency_window_seconds: int
    allow_market_orders: bool
    require_reference_price_for_market_order: bool
    kill_switch_enabled: bool
    configuration_source: str
    effective_at: datetime
    warnings: list[str]


class SignalRiskAssessmentBody(BaseModel):
    account_id: UUID
    quantity: str | None = None
    reference_price: str | None = None
    idempotency_key: str = Field(min_length=1, max_length=128)

    @field_validator("quantity", "reference_price", mode="before")
    @classmethod
    def decimal_strings_only(cls, value: Any) -> Any:
        if value is None or isinstance(value, str):
            return value
        raise ValueError("decimal values must be JSON strings")


class RiskDecisionPageResponse(BaseModel):
    items: list[RiskDecisionResponse]
    page: int
    page_size: int
    total: int
